netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,305 @@
+      pendingStreams.put(streamId, pendingStream);
+    }
+    pendingStream.frames.add(new HeadersFrame(headers, streamDependency, weight, exclusive,
+            padding, endOfStream, promise));
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
+                                      ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeRstStream(ctx, streamId, errorCode, promise);
+    }
+    // Since the delegate doesn't know about any buffered streams
+    // we have to handle cancellation of the promises and releasing of the ByteBufs
+    // here.
+    PendingStream stream = pendingStreams.remove(streamId);
+    if (stream != null) {
+      stream.fail();
+      promise.setSuccess();
+    } else {
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                 boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream != null) {
+      pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+    } else {
+      ReferenceCountUtil.safeRelease(data);
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+    ChannelFuture future = super.writeSettingsAck(ctx, promise);
+    // After having received a SETTINGS frame, the maximum number of concurrent streams
+    // might have changed. So try to create some buffered streams.
+    tryCreatePendingStreams();
+    return future;
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    failPendingStreams();
+  }
+
+  private void tryCreatePendingStreams() {
+    while (!pendingStreams.isEmpty() && connection().local().acceptingNewStreams()) {
+      createNextPendingStream();
+    }
+  }
+
+  private void failPendingStreams() {
+    while (!pendingStreams.isEmpty()) {
+      PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+      stream.fail();
+    }
+  }
+
+  private void createNextPendingStream() {
+    Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+    if (entry == null) {
+      return;
+    }
+    PendingStream pendingStream = entry.getValue();
+    pendingStream.sendFrames();
+    largestCreatedStreamId = pendingStream.streamId;
+  }
+
+  private boolean existingStream(int streamId) {
+    return streamId <= largestCreatedStreamId;
+  }
+
+  private class PendingStream {
```

Comment:
@ejona86 had that at first, but due to some of the design choices that we made during the review it can't be static (need `writeHeaders`, `writeData` methods from the outer class).